### PR TITLE
br: download sst again with no-crypt parameters when met decryption error for downloadSST

### PR DESCRIPTION
**Pull request by joccau**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/32423

Problem Summary:
If the SST file from backup crypt is not been encrypted，user will fail to restore because of failing to download sst.

### What is changed and how it works?

- br should try download SST with no-crypt when met this error in this case.
- Simplify function( import() is too long and too complex to understand)

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Review by fengou1 — APPROVED** [review]
go concurrency, go routines fetch local variables, it possible update on next iterator cause problem

File: br/pkg/backup/client.go [hunk 1]
```diff
@@ -675,9 +662,9 @@ func (bc *Client) fineGrainedBackup(
 			go func(boFork *tikv.Backoffer) {
 				defer wg.Done()
 				for rg := range retry {
+					req.StartKey, req.EndKey = rg.StartKey, rg.EndKey
```
Comment (3pointer) [inline br/pkg/backup/client.go:665]: it seems we need clone `req` here.

Comment (joccau) [reply br/pkg/backup/client.go]: > it seems we need clone `req` here.

Fix it already.

File: br/pkg/backup/client.go [hunk 2]
```diff
@@ -868,7 +833,7 @@ func (bc *Client) handleFineGrained(
 		// Handle responses with the same backoffer.
 		func(resp *backuppb.BackupResponse) error {
 			response, shouldBackoff, err1 :=
-				OnBackupResponse(storeID, bo, backupTS, lockResolver, resp)
+				OnBackupResponse(storeID, bo, req.StartVersion, lockResolver, resp)
```
Comment (3pointer) [inline br/pkg/backup/client.go:836]: should it be `EndVersion`

File: br/pkg/restore/import.go [hunk 3]
```diff
@@ -270,32 +303,17 @@ func (importer *FileImporter) Import(
...
+	startKey, endKey, err := importer.getKeyRangeForFiles(files, rewriteRules)
+	if err != nil {
+		return errors.Trace(err)
 	}
 
-	log.Debug("rewrite file keys",
-		logutil.Files(files),
-		logutil.Key("startKey", startKey),
-		logutil.Key("endKey", endKey))
+	log.Debug("rewrite file keys", logutil.Files(files),
```
Comment (3pointer) [inline br/pkg/restore/import.go:313]: it's better move this log to https://github.com/pingcap/tidb/pull/32847/files#diff-7b2d077bc48a0da7a3532fa8cdce121bf541ea6146dd8b064bdee212a84a7e55R278

**Review by 3pointer — COMMENTED** [review]
rest LGTM

File: br/pkg/backup/client.go (hunk 2 above)
Comment (joccau) [reply br/pkg/backup/client.go]: > should it be `EndVersion`🤔

Fix it already, PTAL again~

File: br/pkg/restore/import.go [hunk 4]
```diff
@@ -259,6 +260,40 @@ func (importer *FileImporter) SetRawRange(startKey, endKey []byte) error {
...
+			start, end, err = rewriteFileKeys(f, rewriteRules)
+			if err != nil {
+				return nil, nil, errors.Trace(err)
+			}
+		}
+
+		if len(startKey) == 0 || bytes.Compare(start, startKey) < 0 {
+			startKey = start
+		}
+		if len(endKey) == 0 || bytes.Compare(startKey, end) < 0 {
```
Comment (YuJuncen) [inline br/pkg/restore/import.go:287]: I guess?
```suggestion
		if len(endKey) == 0 || bytes.Compare(endKey, end) < 0 {
```

**Review by YuJuncen — APPROVED** [review]
rest lgtm

**Pull request merged**